### PR TITLE
Faction traits and carp null rod fix

### PR DESCRIPTION
**Pull request by TheBonded**
## About The Pull Request

- adds a faction trait macro thingy that allows other factions to have related traits without filling up the defines file
- if you have the carp faction from the null rod variant or sleeping carp you will no longer lose the faction by taking off a carp costume
- the null rod carp plushie variant used to fail if you were too holy and work if you weren't holy due to the check being backwards. How did no one notice this until now?

## Why It's Good For The Game

Fixes good

## Changelog

:cl:
fix: You can now gain the carp faction from multiple sources without losing it
fix: The chaplain's null rod carp plushie variant now properly works on holy people instead of only working on the unworthy
/:cl:

### CODE REVIEW

**Comment by TheBonded** [comment]
Tested and ready for review

**Comment by timothymtorres** [comment]
This closes https://github.com/tgstation/tgstation/issues/42724 since we tested it and found no problems.

**Comment by TheBonded** [comment]
> This closes #42724 since we tested it and found no problems.

@timothymtorres that report should be closed regardless of this PR, as our tests didn't change how the carp faction works, so that bug no longer occurs anymore whether this is merged or not

**Comment by ExcessiveUseOfCobblestone** [comment]
If you have to put in a specific string and make sure it’s the same capitalization and spelling across the board what’s the point of the define?

**Comment by timothymtorres** [comment]
> If you have to put in a specific string and make sure it’s the same capitalization and spelling across the board what’s the point of the define?
> 
> 
> 
> almost like you need a second define for the inside lol

It's so that you don't need a define for each faction, the macro will handle it.  This also scales really well since as more factions get added or removed, you don't need to edit the defines to compensate.

**Comment by ExcessiveUseOfCobblestone** [comment]
you do need a define though because the current system requires that every instance of "carp" has to be changed if you decided to change that entry. It's a magic string moreso than just setting "carpfaction" or what have you everywhere.

I can understand the trait macro, but you still need the inside item macro'd as well.

File: code/__DEFINES/traits.dm [hunk 1]
```diff
@@ -349,8 +349,10 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_SPARRING "sparring"
 /// The user is currently challenging an elite mining mob. Prevents him from challenging another until he's either lost or won.
 #define TRAIT_ELITE_CHALLENGER "elite_challenger"
-
-#define TRAIT_NOBLEED "nobleed" //This carbon doesn't bleed
+/// This mob has a faction from a source
+#define TRAIT_FACTION(faction) "faction_[faction]"
```
Comment (optimumtact) [inline code/__DEFINES/traits.dm:353]: this is a goofy fix and you're better off just rolling all factions into traits

**Comment by TheBonded** [comment]
@optimumtact 

@tralezab told me the macro would be a better fix, considering if we don't do the macro then we'll have to have a check in the faction granter that checks SPECIFICALLY for the carp faction, and as more faction traits are added there would be a massive line of 
```
if(faction_to_grant == SPECIES)
      ADD_TRAIT(user, TRAIT_FACTION_SPECIES, TRAIT_SOURCE)
```
for EVERY faction that can be granted by the faction granter. With the macro, each time someone adds a new faction granter they can just use what I have here right now. 

Just like how the faction granter is made to allow you to grant multiple different factions with it, even though it's only used to add the carp faction right now, it still uses the faction_to_grant so it can be used in different places in the future.

**Comment by TheBonded** [comment]
Alright well if anyone wants to replace all factions with traits, please do. Until then, this bug will continue to ravage the servers, mildly inconveniencing the 1/150000 people trying to do a carp gimmick once or twice a year.

**Pull request closed without merging**